thriftbp/server.go

```diff
@@ -91,36 +93,145 @@ func NewServer(cfg ServerConfig) (*thrift.TSimpleServer, error) {
+func NewBaseplateServerFromOpts(
+	bp baseplate.Baseplate,
+	opts ...ServerOpt,
+) (baseplate.Server, error) {
+
+	srv, err := NewServerFromOpts(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return ApplyBaseplate(bp, srv), nil
+}
+
+// ServerOpt is a type used for defining configuration arguments
+// needed for creating a baseplate server. It allows returning
+// an error to signal that a configuration value was invalid
+// or failed to be set.
+type ServerOpt func(cfg *ServerConfig) error
+
+// BuildConfig creates a new ServerConfig instance and applies
+// all the supplied configuration options to it.
+func BuildConfig(opts ...ServerOpt) (*ServerConfig, error) {
+	cfg := &ServerConfig{}
+	for _, opt := range opts {
+		err := opt(cfg)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cfg, nil
+}
+
+// ServerOptFrom returns a server option that overwrites the
+// any existing values in a ServerConfig with those defined
+// in src.
+func ServerOptFrom(src ServerConfig) ServerOpt {
+	return func(dst *ServerConfig) error {
+		*dst = src
+		return nil
+	}
+}
+
+// DefaultServerOpts builds and returns a slice of the default
+// baseplate server config options.
+func DefaultServerOpts(bp baseplate.Baseplate) []ServerOpt {
+	return []ServerOpt{
+		serverOptMiddleware(bp),
+		serverOptLogger(bp),
+		serverOptAddress(bp),
+		serverOptSocket(bp),
+	}
+}
+
+func serverOptMiddleware(bp baseplate.Baseplate) ServerOpt {
+	return func(cfg *ServerConfig) error {
+		middlewares := BaseplateDefaultProcessorMiddlewares(
+			DefaultProcessorMiddlewaresArgs{
+				EdgeContextImpl:                    bp.EdgeContextImpl(),
+				ErrorSpanSuppressor:                cfg.ErrorSpanSuppressor,
+				ReportPayloadSizeMetricsSampleRate: cfg.ReportPayloadSizeMetricsSampleRate,
+			},
+		)
+		cfg.Middlewares = append(middlewares, cfg.Middlewares...)
+
+		return nil
+	}
+}
+
+func serverOptLogger(bp baseplate.Baseplate) ServerOpt {
+	return func(cfg *ServerConfig) error {
 		cfg.Logger = log.ZapWrapper(log.ZapWrapperArgs{
 			Level: bp.GetConfig().Log.Level,
 			KVPairs: map[string]interface{}{
 				"from": "thrift",
 			},
 		}).ToThriftLogger()
+
+		return nil
 	}
-	cfg.Addr = bp.GetConfig().Addr
-	cfg.Socket = nil
-	srv, err := NewServer(cfg)
-	if err != nil {
-		return nil, err
+}
+
+// serverOptAddress is used to copy the address config from the
+// baseplate config to the server config. Currently, it's only
+// needed to guarantee backwards compatibility in NewServer
+func serverOptAddress(bp baseplate.Baseplate) ServerOpt {
+	return func(cfg *ServerConfig) error {
+		cfg.Addr = bp.GetConfig().Addr
+
+		return nil
+	}
+}
+
+func serverOptSocket(bp baseplate.Baseplate) ServerOpt {
+	return ServerOptSocket(bp.GetConfig().Addr)
+}
+
+// ServerOptSocket creates a server option that initializes a
+// thrift.TServerSocket listening on the address supplied.
+func ServerOptSocket(address string) ServerOpt {
```

Comment:
ServerOptSocket should take a socket.  A name like `WithListenAddr` would be more appropriate for this option